### PR TITLE
Spike: lazy union reader

### PR DESCRIPTION
**Pull request by willmurphyscode**
## TL;DR

The elf-binary-package-cataloger sometimes gets OOM-killed, and generally is responsible for a lot of allocations. This spike adds a lazy union reader which avoids some unnecessary calls to `io.ReadAll`. The code is working, at least as far as the unit tests added and some manual testing an profiling, but probably wants some clean up and more careful testing.

Rough breakdown of profiling:
1. Running on `main` with all default image catalogers on this test image allocates ~69.53 GB over the course of the run
2. Running on this branch with the same setup allocates about 64.05 GB, for a savings of ~5.5 GB allocations
3. Running on `main` with the elf-binary-package-cataloger turned off allocates ~47.65 GB, or a savings of about 22 GB.

### Questions and concerns

1. Are we interested in this approach? Do the savings (reducing total allocations by elf cataloger by ~25%) justify the risks+costs of maintaining our own little reader package?
4. Are there other investigations that should be performed?
5. I'm a bit concerned with a homemade implementation in such a central place. Specifically, there are `int` and `int64` in the interfaces around io reads, and it would be easy to have an overflow or other bug. Any suggestions for missing test cases?

## Details

This is a spike to try to reduce some of Syft's memory allocations; in particular we have had user reports of OOM when running the `elf-binary-package-cataloger`. ([report](https://github.com/anchore/syft/issues/2704#issuecomment-2040387787), [report](https://anchorecommunity.slack.com/archives/C019BUXV7R6/p1713732067638659)).

Currently, Syft uses interfaces to hide the underlying file system from the catalogers, because the filesystem might be a tar, or a representation of an OCI image, or a directory, or whatever, and the catalogers shouldn't need to know about those differences. In order to achieve this, Syft uses a type called `UnionReader`: https://github.com/anchore/syft/blob/5326efcef13db1df2f03e77ce580feef9857569d/syft/internal/unionreader/union_reader.go#L14-L19

Currently, instantiating a UnionReader, if the underlying reader doesn't implement ReadAt or Seek, results in calling `io.ReadAll` on the underlying reader, and using the resulting bytes to back the Seek and ReadAt calls: https://github.com/anchore/syft/blob/5326efcef13db1df2f03e77ce580feef9857569d/syft/internal/unionreader/union_reader.go#L47

The new `elf-binary-package-cataloger` needs to call ReadAt and Seek on a potentially large number of files, resulting in a lot of `io.ReadAll` calls if this cataloger is enabled.

This branch spikes out making a new struct called `lazyUnionReader`, which wraps an `io.ReadCloser` and presents a `UnionReader`. Basically, instead of calling `io.ReadAll` when it is created, it copies all bytes read (or `Seek`ed past) into a backing byte slice. In other words, it has similar behavior as the current UnionReader, except that the allocation is lazy; if we never read past byte 102 in a given file, the backing array will only be (about) 102 bytes long, instead of the whole length of the file.

I thought maybe another package in the go ecosystem would provide a lazy read seeker, but I haven't found one, so this branch implements one with some unit tests.

I'll post comments with some profiling.

### CODE REVIEW

**Comment by willmurphyscode** [comment]
Here are 3 profiles:

## Main, all default image catalogers:
image

## this branch, all default image catalogers:
image

## main, `elf-binary-package-cataloger` specifically disabled:
image

**Comment by willmurphyscode** [comment]
The profiles were generated by adding the following code (+ necessary imports) to `cmd/syft/main.go`:

``` go
	defer func() {
		f, err := os.Create("mem.prof")
		defer f.Close()
		if err != nil {
			panic(err)
		}
		runtime.GC() // get up to data mem info
		if err = pprof.WriteHeapProfile(f); err != nil {
			panic(err)
		}
	}()
```

And then rendered by running `go tool pprof -http=:8080 mem.prof`, selecting "Flame graph" from the view menu, and selecting "alloc_space" from the sample menu.

**Comment by willmurphyscode** [comment]
@kzantow asks a good question: How much do we save of maximum heap size, rather than total allocations over the whole execution?

I'll try to find the answer, and profile some more images.

**Comment by willmurphyscode** [comment]
Superseded by #2814
